README
Resilient Image Library (ResIL) version 1.7.9 Readme, Notes and Quick Use
-------------------------------------------------------------------------

<DZA[afk]> ResIL song: "la la la, a cross-platform image library utilizing a
           simple syntax to load, save, convert, manipulate, filter and display
           a variety of images with ease, la la la"


What is it?
-----------

ResIL is an Open Source image library whose distribution is done under the
terms of the GNU LGPL license. See the COPYING file for more details.
It was created as an updated fork of DevIL. Like DevIL, ResIL offers you a 
simple way to implement loading, manipulating, filtering, converting, 
displaying, saving from/to several different image formats in your own project.

ResIL source code uses code from the following projects:

ILMBase
Jasper
The Independent JPEG Group's JPEG software
Little CMS
libMNG
libPNG
OpenEXR
Tiff
zlib

Please refer to the idividual license files contained in the full source of ResIL
for the respective licenses of the projects.


Where can I find it?
--------------------

ResIL can be found at http://sourceforge.net/projects/resil


How do I build and install the 3 libraries ?
-----------------------------------------

*nix    users should read README.unix
VisualC users should read README.win
Cygwin  users should read README.cygwin
MacOSX  users should read README.macosx

PS: *nix stands for GNU/Linux, *BSD, SunOS/Solaris and perhaps some more.
 

More Extensive Documentation
----------------------------

This file is only a quick guide to point you to more detailed information on
how to use ResIL.  More extensive documentation can currently be found on the
ResIL site at http://openil.sf.net and in the /Docs directory in a normal
install.


Why the hell another image library?
-----------------------------------

I have never seen an image library that can do everything ResIL does.  Sure,
various different libraries can do part of what ResIL can do as well or even
better, but I wanted a simple to use library that encompassed all of these
features.  I also wanted an extremely portable image library that could be used
from a variety of languages and utilized the OpenGL syntax.


Basic Readme
------------

Most anything stated in this document applies to ResIL as well as ResILU and
ResILUT, unless otherwise stated. (This file is best viewed with word wrap on.)


Errors:
-------

All errors generated inside ResIL, along with illegal parameters passed to
ResIL functions are caught and passed to ilSetError(), an internal library
function.  The calling program can call ilGetError() to get the value of the
error generated.  Error types are defined in il.h, using the 0x501 - 0x5FF
range.  ilGetError() will return 0 (IL_NO_ERROR) if no error has occurred.


Basic Usage:
------

This demonstrates loading an image through ResIL for OpenGL. Don't forget to 
call ilInit before you before you do anything:

#include <IL/il.h>
#include <IL/ilu.h>
#include <IL/ilut.h>

...

ILuint ResILError;


ilInit();

ResILError = ilGetError();

if (ResILError != IL_NO_ERROR) {
  printf ("ResIL Error (ilInit: %s\n", iluGetErrorString (ResILError));
  exit (2);
}

....

ILuint ResILID;


ilGenImages(1, &ResILID);
ilBindImage(ResILID);
ilLoadImage("default1.tga");  // Loads into the current bound image
ResILError = ilGetError();

if (ResILError != IL_NO_ERROR) {
  printf ("ResIL Error (ilLoadImage: %s\n", iluGetErrorString (ResILError));
  exit (2);
}

....

ilutRenderer(IL_OPENGL);  // Switch the renderer

....

GLuint openglID, openglError;


openglID   = ilutGLBindTexImage(); // This generates the texture for you
ResILError = ilGetError();

if (ResILError != IL_NO_ERROR) {
  printf ("Error: %s\n", iluGetErrorString (ResILError));
  exit (2);
}

if (openglError != GL_NO_ERROR) {
  printf ("Opengl Error (ilutGLBindTexImage): %s\n", gluGetErrorString (openglError));
  exit (2);
}



// Make sure to close the image when you are done with it (though ResIL
// automatically deletes them when the program exits):

glDeleteTextures(1, &openglID);
ilDeleteImages  (1, &ResILID);


More Examples:
---------

The TestIL project is included to test features of ResIL.

ResIL includes a project called GLTest.  This is a simple test of ResIL's
capabilities.  All it does it load any image and displays it in a window
created by FreeGlut, which is available on http://freeglut.sourceforge.net. It
is also included to let the user have an idea of what the library can really
be used for.

Several other test projects are included to test support with various display
APIs.  The WindowsTest project is a basic image program that only runs in
Windows right now but showcases several of ResIL's features through various
menus.

If you want more in-depth tutorials, you can find them on
http://openil.sf.net, or they may be in your installation under the /examples
directory.  Documents are also available in the /docs directory.


Additional Reading
------------------

All image formats used in ResIL have corresponding documents on
http://www.wotsit.org, under the Graphics Files section.  These documents
proved invaluable for the creation of this library when there was no library
already available for that image format.


Legalese
--------

All contents of this file are intellectual property of Denton Woods,
copyright 2001-2008. Last edited by Björn Ganster.


ChangeLog
ResIL 1.8.3
	* Added new thread-safe API 2.0 (work in progress, unstable!)
		* Old API 1.x still exposed via il.h
		* API 2.x exposed via il2.h
		* ResIL uses API 2.x internally
	* Various fixes for Linux support contributed by Bernhard Rosenkränzer
	* Uses libpng 1.6.12
	* Uses lcms 2.6
	* Fixes for file formats

2014-05-08 Björn Ganster <mail@bjoern-ganster.de>, ResIL 1.8.2
	* File format fixes: cut, pix, hdr, sun ras, vtf?
	* Each image now has its own set of persistence functions

2014-04-11 Björn Ganster <mail@bjoern-ganster.de>, ResIL 1.8.1
	* New function: ilLoadFuncs reads a file using the read functions defined in ilSetRead
	* New function: ilSaveFuncs saves a file using the write functions defined in ilSetWrite
	* Improved performance for file type detection
	* Stability fixes for GIF, BMP
	* Work towards eliminating global variables
	* A collection of automated tests for loading and saving files
	* Some redundant code removed

2014-03-21 Björn Ganster <mail@bjoern-ganster.de>, ResIL 1.8.0
	* Fixed a bug in il_gif.cpp that caused a heap corruption
	  when the index for the transparent color pointed beyond
	  the current palette
	* Converted several files to cpp
	
2014-03-17 Björn Ganster <mail@bjoern-ganster.de>
	* Fixed a bug for .ico files that contain png data that 
	  caused a NULL pointer exception when no mask was specified

2009-03-08 Matěj Týč <bubla@users.sf.net>
	* docs/Makefile.am: Fixed distcheck by overriding dvi target.
	* lib/Makefile.am: Removed ilu_mipmap.h file
	* configure.ac: Added option to disable checks for 'nvidia texture tools' and
	'libsquish'

2009-03-08 Denton Woods <doomwiz@users.sf.net>
	* il_ilbm.c:410: Changed to IL_BGR from IL_RGB.
	* il_iff.c (ilLoadIffF/ilLoadIffL): Changed to call Lbm loading functions
	  if Iff loading fails, since they both can have the .iff extension.
	* il_ilbm.c:416: Added origin reference.
	* configure.ac: Added more file formats to the format checks.
	* Makefile.am: Removed ilu_mipmap.h.  Added ilu_error-french.h.

2009-03-07 Denton Woods <doomwiz@users.sf.net>
	* Updated version numbers to 1.7.8 project-wide.
	* DevIL_manual.texi: Added several file formats to the appendix.
	* il_io.c: Updated Doxygen comments for loading/saving functions.
	* libraries.txt: Updated with new libraries.
	* il_io.c (ilTypeFromExt): Changed return value of FITS files to IL_FITS.
		Added DICOM, FTX, Rot and Texture checks.
	* il_states.c/.h: Updated with newer file formats.
	* il_blp.c:649: Fixed ilNewImageFull call to use color index.
			   659: Fixed memcpy from base image to new mipmap.
	* il_dcx.c, il_exr.cpp and many more: Checks return value of ilFixImage now.
	* il_iff.c:271: Got rid of ilConvertImage call.

2009-03-06 Denton Woods <doomwiz@users.sf.net>
	* Updated VC8 projects to use il_mp3.c.
	* il.h: Added description to IL_WDP.

2009-03-05 Denton Woods <doomwiz@users.sf.net>
	* il_mp3.c: Added this file.
	* il_io.c: Added loading calls for .mp3 files.
	* il_internal.h: Added .mp3 loading declarations.
	* Makefile.am: Added il_mp3.c to source line.
	* il.h: Added IL_MP3 define.

2009-03-04 Denton Woods <doomwiz@users.sf.net>
	* il_utx.cpp: Started cleanup of file using C++ constructs.
	* Updated VC8 projects to use il_utx.cpp.

2009-03-03 Denton Woods <doomwiz@users.sf.net>
	* ilu_error-french.h: Added French translation of errors.
	* ilu.h: Added ILU_FRENCH define.
	* Changed il_utx.c to il_utx.cpp.

2009-03-02 Denton Woods <doomwiz@users.sf.net>
	* il_utx.c: Added DXT1 loading in UTX files.
	* il_utx.h: Added this file.
	* lib/Makefile.am: Added il_utx.h to include line.
	* il_utx.c: Fixed loading of larger images.
	* windows_example.rc: Fixed accelerator for menu.

2009-03-01 Denton Woods <doomwiz@users.sf.net>
	* il_utx.c: Added this file.
	* lib/Makefile.am: Added il_utx.c to source line.
	* il_internal.h: Added .utx loading declarations.
	* il_io.c: Added loading calls for .utx files.
	
2009-02-28 Denton Woods <doomwiz@users.sf.net>
	* il_states.c / il_states.h: Added ilVtfCompression to IL_STATES.
	* il_vtf.c: Added more formats to saving.
	* il.h: Added IL_VTF_COMP define.
	* il_io.c: Added saving calls for .vtf files.
	* il_dds.c: Added DecompressARGB16 to load a2r10g10b10 and a2b10g10r10 formats.
	* il_convbuff (iSwitchTypes): Added several missing break statements.
	* il_convbuff:1988: Changed to check types instead of bpc.

2009-02-27 Denton Woods <doomwiz@users.sf.net>
	* il_dpx.c / il_dpx.h: Got basic dpx loading working.
	* WindowsTest.cpp: Fixed memory leak when loading files.
	* il_vtf.c (iGetVtfHead): Added support for headers of 64 bytes.
	* il_iwi.c (iLoadIwiInternal): Added IL_INVALID_FILE_HEADER error.
	* il_vtf.c: Added VTF writing functions.

2009-02-26 Denton Woods <doomwiz@users.sf.net>
	* configure.ac, lib/Makefile.am, il.h, il_internal.h, il_ilbm.c, il_io.c: Added
	  patch from Ben Campbell for ILBM files at
	  https://sourceforge.net/tracker2/index.php?func=detail&aid=2635333&group_id=4470&atid=304470.
	* il_dpx.c / il_dpx.h: Added these files.
	* lib/Makefile.am: Added il_dpx.c and il_dpx.h to source/include lines.
	* il_internal.h: Added .dpx loading declarations.
	* il_io.c: Added loading calls for .dpx files.
		
2009-02-25 Denton Woods <doomwiz@users.sf.net>
	* il_iwi.c:329: Changed IL_MAX parameter from 16 to 8 (copy+paste error).
	* WindowsTest.cpp: Added mipmap functionality to 0-9 keys (0 goes to main image).
	
2009-02-24 Denton Woods <doomwiz@users.sf.net>
	* vc9 Unicode projects: Changed settings to use x64 and x86 directories for .obj files.
	* test/in: Removed this directory.
	* Added "Test Images" to SVN.
	* il_dds.c (DecompressFloat): Added proper support for G16B16, G32B32, R16 and R32 images.

2009-02-23 Denton Woods <doomwiz@users.sf.net>
	* il_pnm.c: Fixed bug at https://sourceforge.net/forum/message.php?msg_id=6491617 -
	  Changed order of lines 448 and 450.

2009-02-22 Denton Woods <doomwiz@users.sf.net>
	* il_internal.h: Added IL_NO_EXTLIBS define.

src-IL/src/il_dcx.cpp
//-----------------------------------------------------------------------------
//
// ImageLib Sources
// Copyright (C) 2000-2009 by Denton Woods
// Last modified: 03/07/2009
//
// Filename: src-IL/src/il_dcx.c
//
// Description: Reads from a .dcx file.
//
//-----------------------------------------------------------------------------


#include "il_internal.h"
#ifndef IL_NO_DCX
#include "il_manip.h"


#ifdef _WIN32
#pragma pack(push, packed_struct, 1)
#endif
typedef struct DCXHEAD
{
	ILubyte		Manufacturer;
	ILubyte		Version;
	ILubyte		Encoding;
	ILubyte		Bpp;
	ILushort	Xmin, Ymin, Xmax, Ymax;
	ILushort	HDpi;
	ILushort	VDpi;
	ILubyte		ColMap[48];
	ILubyte		Reserved;
	ILubyte		NumPlanes;
	ILushort	Bps;
	ILushort	PaletteInfo;
	ILushort	HScreenSize;
	ILushort	VScreenSize;
	ILubyte		Filler[54];
} IL_PACKSTRUCT DCXHEAD;
#ifdef _WIN32
#pragma pack(pop, packed_struct)
#endif


// Internal function obtain the .dcx header from the current file.
ILboolean iGetDcxHead(SIO* io, DCXHEAD *Head)
{
	Head->Xmin = GetLittleUShort(io);
	Head->Ymin = GetLittleUShort(io);
	Head->Xmax = GetLittleUShort(io);
	Head->Ymax = GetLittleUShort(io);
	Head->HDpi = GetLittleUShort(io);
	Head->VDpi = GetLittleUShort(io);
	Head->Bps = GetLittleUShort(io);
	Head->PaletteInfo = GetLittleUShort(io);
	Head->HScreenSize = GetLittleUShort(io);
	Head->VScreenSize = GetLittleUShort(io);

	return IL_TRUE;
}


// Internal function to get the header and check it.
ILboolean iIsValidDcx(SIO* io)
{
	ILuint Signature;

	if (io->read(io, &Signature, 1, 4) != 4)
		return IL_FALSE;
	io->seek(io, -4, IL_SEEK_CUR);

	return (Signature == 987654321);
}


// Internal function used to check if the HEADER is a valid .dcx header.
// Should we also do a check on Header->Bpp?
ILboolean iCheckDcx(DCXHEAD *Header)
{
	ILuint	Test, i;

	// There are other versions, but I am not supporting them as of yet.
	//	Got rid of the Reserved check, because I've seen some .dcx files with invalid values in it.
	if (Header->Manufacturer != 10 || Header->Version != 5 || Header->Encoding != 1/* || Header->Reserved != 0*/)
		return IL_FALSE;

	// See if the padding size is correct
	Test = Header->Xmax - Header->Xmin + 1;
	/*if (Header->Bpp >= 8) {
		if (Test & 1) {
			if (Header->Bps != Test + 1)
				return IL_FALSE;
		}
		else {
			if (Header->Bps != Test)  // No padding
				return IL_FALSE;
		}
	}*/

	for (i = 0; i < 54; i++) {
		if (Header->Filler[i] != 0)
			return IL_FALSE;
	}

	return IL_TRUE;
}


ILimage *iUncompressDcxSmall(ILimage* image, DCXHEAD *Header)
{
	ILuint	i = 0, j, k, c, d, x, y, Bps;
	ILubyte	HeadByte, Colour, Data = 0, *ScanLine = NULL;
	ILimage	*Image;
	SIO* io = &image->io;

	Image = il2NewImage(Header->Xmax - Header->Xmin + 1, Header->Ymax - Header->Ymin + 1, 1, Header->NumPlanes, 1);
	if (Image == NULL)
		return NULL;

	/*if (!ilTexImage(Header->Xmax - Header->Xmin + 1, Header->Ymax - Header->Ymin + 1, 1, 1, 0, IL_UNSIGNED_BYTE, NULL)) {
		return IL_FALSE;
	}*/
	Image->Origin = IL_ORIGIN_UPPER_LEFT;

	switch (Header->NumPlanes)
	{
		case 1:
			Image->Format = IL_LUMINANCE;
			break;
		case 4:
			Image->Format = IL_COLOUR_INDEX;
			break;
		default:
			il2SetError(IL_ILLEGAL_FILE_VALUE);
			ilCloseImage(Image);
			return NULL;
	}

	if (Header->NumPlanes == 1) {
		for (j = 0; j < Image->Height; j++) {
			i = 0;
			while (i < Image->Width) {
				if (io->read(io, &HeadByte, 1, 1) != 1)
					goto file_read_error;
				if (HeadByte >= 192) {
					HeadByte -= 192;
					if (io->read(io, &Data, 1, 1) != 1)
						goto file_read_error;

					for (c = 0; c < HeadByte; c++) {
						k = 128;
						for (d = 0; d < 8 && i < Image->Width; d++) {
							Image->Data[j * Image->Width + i++] = (!!(Data & k) == 1 ? 255 : 0);
							k >>= 1;
						}
					}
				}
				else {
					k = 128;
					for (c = 0; c < 8 && i < Image->Width; c++) {
						Image->Data[j * Image->Width + i++] = (!!(HeadByte & k) == 1 ? 255 : 0);
						k >>= 1;
					}
				}
			}
			if (Data != 0)
				io->getc(io);  // Skip pad byte if last byte not a 0
		}
	}
	else {   // 4-bit images
		Bps = Header->Bps * Header->NumPlanes * 2;
		Image->Pal.Palette = (ILubyte*)ialloc(16 * 3);  // Size of palette always (48 bytes).
		Image->Pal.PalSize = 16 * 3;
		Image->Pal.PalType = IL_PAL_RGB24;
		ScanLine = (ILubyte*)ialloc(Bps);
		if (Image->Pal.Palette == NULL || ScanLine == NULL) {
			ifree(ScanLine);
			ilCloseImage(Image);
			return NULL;
		}

		memcpy(Image->Pal.Palette, Header->ColMap, 16 * 3);
		imemclear(Image->Data, Image->SizeOfData);  // Since we do a += later.

		for (y = 0; y < Image->Height; y++) {
			for (c = 0; c < Header->NumPlanes; c++) {
				x = 0;
				while (x < Bps) {
					if (io->read(io, &HeadByte, 1, 1) != 1)
						goto file_read_error;
					if ((HeadByte & 0xC0) == 0xC0) {
						HeadByte &= 0x3F;
						if (io->read(io, &Colour, 1, 1) != 1)
							goto file_read_error;
						for (i = 0; i < HeadByte; i++) {
							k = 128;
							for (j = 0; j < 8; j++) {
								ScanLine[x++] = !!(Colour & k);
								k >>= 1;
							}
						}
					}
					else {
						k = 128;
						for (j = 0; j < 8; j++) {
							ScanLine[x++] = !!(HeadByte & k);
							k >>= 1;
						}
					}
				}

				for (x = 0; x < Image->Width; x++) {  // 'Cleverly' ignores the pad bytes. ;)
					Image->Data[y * Image->Width + x] += ScanLine[x] << c;
				}
			}
		}
		ifree(ScanLine);
	}

	return Image;

file_read_error:
	ifree(ScanLine);
	ilCloseImage(Image);
	return NULL;
}


// Internal function to uncompress the .dcx (all .dcx files are rle compressed)
ILimage *iUncompressDcx(ILimage* image, DCXHEAD *Header)
{
	ILubyte		ByteHead, Colour, *ScanLine = NULL /* Only one plane */;
	ILuint		c, i, x, y;//, Read = 0;
	ILimage		*Image = NULL;
	SIO* io = &image->io;

	if (Header->Bpp < 8) {
		/*il2SetError(IL_FORMAT_NOT_SUPPORTED);
		return IL_FALSE;*/
		return iUncompressDcxSmall(image, Header);
	}

	Image = il2NewImage(Header->Xmax - Header->Xmin + 1, Header->Ymax - Header->Ymin + 1, 1, Header->NumPlanes, 1);
	if (Image == NULL)
		return NULL;
	/*if (!ilTexImage(Header->Xmax - Header->Xmin + 1, Header->Ymax - Header->Ymin + 1, 1, Header->NumPlanes, 0, IL_UNSIGNED_BYTE, NULL)) {
		return IL_FALSE;
	}*/
	Image->Origin = IL_ORIGIN_UPPER_LEFT;

	ScanLine = (ILubyte*)ialloc(Header->Bps);
	if (ScanLine == NULL)
		goto dcx_error;

	switch (Image->Bpp)
	{
		case 1:
			Image->Format = IL_COLOUR_INDEX;
			Image->Pal.PalType = IL_PAL_RGB24;
			Image->Pal.PalSize = 256 * 3; // Need to find out for sure...
			Image->Pal.Palette = (ILubyte*)ialloc(Image->Pal.PalSize);
			if (Image->Pal.Palette == NULL)
				goto dcx_error;
			break;
		//case 2:  // No 16-bit images in the dcx format!
		case 3:
			Image->Format = IL_RGB;
			Image->Pal.Palette = NULL;
			Image->Pal.PalSize = 0;
			Image->Pal.PalType = IL_PAL_NONE;
			break;
		case 4:
			Image->Format = IL_RGBA;
			Image->Pal.Palette = NULL;
			Image->Pal.PalSize = 0;
			Image->Pal.PalType = IL_PAL_NONE;
			break;

		default:
			il2SetError(IL_ILLEGAL_FILE_VALUE);
			goto dcx_error;
	}

	//TODO: because the .pcx-code was broken this
	//code is probably broken, too
	for (y = 0; y < Image->Height; y++) {
		for (c = 0; c < Image->Bpp; c++) {
			x = 0;
			while (x < Header->Bps) {
				if (io->read(io, &ByteHead, 1, 1) != 1) {
					goto dcx_error;
				}
				if ((ByteHead & 0xC0) == 0xC0) {
					ByteHead &= 0x3F;
					if (io->read(io, &Colour, 1, 1) != 1) {
						goto dcx_error;
					}
					for (i = 0; i < ByteHead; i++) {
						ScanLine[x++] = Colour;
					}
				}
				else {
					ScanLine[x++] = ByteHead;
				}
			}

			for (x = 0; x < Image->Width; x++) {  // 'Cleverly' ignores the pad bytes ;)
				Image->Data[y * Image->Bps + x * Image->Bpp + c] = ScanLine[x];
			}
		}
	}

	ifree(ScanLine);

	// Read in the palette
	if (Image->Bpp == 1) {
		ByteHead = io->getc(io);	// the value 12, because it signals there's a palette for some reason...
							//	We should do a check to make certain it's 12...
		if (ByteHead != 12)
			io->seek(io, -1, IL_SEEK_CUR);
		if (io->read(io, Image->Pal.Palette, 1, Image->Pal.PalSize) != Image->Pal.PalSize) {
			ilCloseImage(Image);
			return NULL;
		}
	}

	return Image;

dcx_error:
	ifree(ScanLine);
	ilCloseImage(Image);
	return NULL;
}


// Internal function used to load the .dcx.
ILboolean iLoadDcxInternal(ILimage* image)
{
	DCXHEAD	Header;
	ILuint	Signature, i, Entries[1024], Num = 0;
	ILimage	*Image, *Base;
	SIO* io = &image->io;

	if (image == NULL) {
		il2SetError(IL_ILLEGAL_OPERATION);
		return IL_FALSE;
	}

	if (!iIsValidDcx(io))
		return IL_FALSE;
	io->read(io, &Signature, 1, 4);

	do {
		if (io->read(io, &Entries[Num], 1, 4) != 4)
			return IL_FALSE;
		Num++;
	} while (Entries[Num-1] != 0);

	for (i = 0; i < Num; i++) {
		io->seek(io, Entries[i], IL_SEEK_SET);
		iGetDcxHead(io, &Header);
		/*if (!iCheckDcx(&Header)) {
			il2SetError(IL_INVALID_FILE_HEADER);
			return IL_FALSE;
		}*/

		Image = iUncompressDcx(image, &Header);
		if (Image == NULL)
			return IL_FALSE;

		if (i == 0) {
			il2TexImage(Image, Image->Width, Image->Height, 1, Image->Bpp, Image->Format, Image->Type, Image->Data);
			Base = image;
			Base->Origin = IL_ORIGIN_UPPER_LEFT;
			ilCloseImage(Image);
		}
		else {
			image->Next = Image;
			image = image->Next;
		}
	}

	return il2FixImage(image);
}


#endif//IL_NO_DCX
